cmd/hypper/upgrade_test.go

```diff
@@ -0,0 +1,423 @@
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package main
+
+import (
+	"fmt"
+	"github.com/rancher-sandbox/hypper/internal/test/ensure"
+	"helm.sh/helm/v3/pkg/action"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"strings"
+	"testing"
+
+	"helm.sh/helm/v3/pkg/chart"
+	"helm.sh/helm/v3/pkg/chart/loader"
+	"helm.sh/helm/v3/pkg/chartutil"
+	"helm.sh/helm/v3/pkg/release"
+)
+
+// FIXME(itxaka): As far as I can see this is done in here so we can  fake the timestamp in the
+// results so they match the golden files
+// We already have this on hypper.go file and I would expect this to be automatically done as we load that file
+// in the case of helm it works exactly like that, their upgrade_test.go does not set the action.Timestamper, its
+// done on the helm_test.go file
+// I have no idea why, but if I dont change the Timestamper then it doesnt match the golden files because it
+// uses the default Time.Now() timestamper...
+func init() {
+	action.Timestamper = testTimestamper
+}
+
+func TestUpgradeCmd(t *testing.T) {
+	tmpChart := ensure.TempDir(t)
+	cfile := &chart.Chart{
+		Metadata: &chart.Metadata{
+			APIVersion:  chart.APIVersionV1,
+			Name:        "funny-bunny",
+			Description: "A Helm chart for Kubernetes",
+			Version:     "0.1.0",
+		},
+	}
+	chartPath := filepath.Join(tmpChart, cfile.Metadata.Name)
+	if err := chartutil.SaveDir(cfile, tmpChart); err != nil {
+		t.Fatalf("Error creating chart for upgrade: %v", err)
+	}
+	ch, err := loader.Load(chartPath)
+	if err != nil {
+		t.Fatalf("Error loading chart: %v", err)
+	}
+	_ = release.Mock(&release.MockReleaseOptions{
+		Name:  "funny-bunny",
+		Chart: ch,
+	})
+
+	// update chart version
+	cfile.Metadata.Version = "0.1.2"
+
+	if err := chartutil.SaveDir(cfile, tmpChart); err != nil {
+		t.Fatalf("Error creating chart: %v", err)
+	}
+	ch, err = loader.Load(chartPath)
+	if err != nil {
+		t.Fatalf("Error loading updated chart: %v", err)
+	}
+
+	// update chart version again
+	cfile.Metadata.Version = "0.1.3"
+
+	if err := chartutil.SaveDir(cfile, tmpChart); err != nil {
+		t.Fatalf("Error creating chart: %v", err)
+	}
+	var ch2 *chart.Chart
+	ch2, err = loader.Load(chartPath)
+	if err != nil {
+		t.Fatalf("Error loading updated chart: %v", err)
+	}
+
+	missingDepsPath := "testdata/testcharts/chart-missing-deps"
+	badDepsPath := "testdata/testcharts/chart-bad-requirements"
+
+	relWithStatusMock := func(n string, v int, ch *chart.Chart, status release.Status) *release.Release {
+		return release.Mock(&release.MockReleaseOptions{Name: n, Version: v, Chart: ch, Status: status})
+	}
+
+	relMock := func(n string, v int, ch *chart.Chart) *release.Release {
+		return release.Mock(&release.MockReleaseOptions{Name: n, Version: v, Chart: ch})
+	}
+
+	tests := []cmdTestCase{
```

Comment:
I suppose in the future (not on this PR) we need to start adding more tests for hypper charts in all commands and subcomands, testing the upgrade etc when there's shared deps, and so on :grimacing:.

Reply:
:+1: 

Agreed, once this settles we need to start adding all types of tests to cover for more.